JSON-RPC: Add finalized and safe blocks

mkalinin opened the pull request:
Adds `finalized` value to the block tags enum.

### Pre-Merge response
There are three options:
* Refer to a block that is likely to stay in the PoW canonical chain forever, i.e. a block which is e.g. `128` blocks behind the head
* Respond with error
* Respond with genesis block

I am slightly inclined towards the third option. It gives users a graceful option to check whether the Merge has happened or not, i.e. checking if `finalized.number` > `earliest.number`.

Whatever option we choose it needs to be specified somewhere. Probably in the description field to this enum.

### Safe block
Another couple of values that are considered for addition to this enum is `safe` and `unsafe`. It worth noting that if these values and the corresponding functionality will be added after the Merge then `unsafe` tag will be used as an alias to the `latest` to avoid backwards incompatibility.

### Description
I've been playing with the way of adding a "description" field to this enum in the openrpc playground. To be honest, it looks ugly as it seems there is no way to describe each value separately and there is also no way to include line breaks into the description. So, whenever we will have to utilize description it will be a sequence of sentences without breaks.

cc @timbeiko @lightclient @djrtwo @MicahZoltu @MariusVanDerWijden

MicahZoltu commented on src/schemas/block.json:
```diff
@@ -125,7 +125,7 @@
 		"title": "Block tag",
 		"type": "string",
-		"enum": ["earliest", "latest", "pending"]
+		"enum": ["earliest", "latest", "pending", "finalized"]
```
```suggestion
		"enum": ["earliest", "latest", "pending", "finalized", "safe", "unsafe"]
```
I would much rather see all of them added sooner rather than later.  Delaying adding the terms doesn't give us any benefit.

I also think that we should add all of these before the merge.  If it were up to me, pre-merge they would map as follows:
```
safe = latest - 1
unsafe = latest
finalized = latest - 256
```

The idea here is to get people used to `safe` lagging a bit behind, but make it useful pre-merge so dapps can start integrating with it.  `unsafe` would map to `latest` as that is how it will map post-merge.  `finalized` being 256 blocks back (about an hour) feels reasonable to me as the timeframe is similar order of magnitude as it will be post-merge and 256 blocks back is *incredibly* unlikely to reorg.

mkalinin replied:
IMO, we shouldn't add `safe` until we have safe block feature fully implemented. `latest - n` may not always be safe which makes `safe` unreliable and potentially dangerous.

Post-merge services will have to use `finalized` instead of `k` confirmations rule post-Merge. But before the merge values for `k` parameter vary, and I am not sure if `k = 256` is used anywhere. We may set `finalized = latest - 128` but there are services that use 16 confirmations and so forth. It makes me think that aliasing `finalized` to k-confirmation rule prior the Merge won't be useful enough for service to make a switch from one to another. I think that `finalized` should be responded with error prior the Merge, it allows to avoid any misuses of this block tag prior the Merge.

We may add `unsafe` with `latest == unsafe`, though, I am not sure if it makes sense if `safe` isn't added.

MicahZoltu replied:
My problem with waiting until post merge to add these is that it leaves us in a situation where dapps cannot deploy merge-ready prior to sometime *after* the merge.  This means that there is a period of time where apps are still using `latest` or `latest - 6` even though the merge is complete and they *shouldn't* be using these values.  I can appreciate that setting these values are somewhat arbitrary prior to the merge, but I think that is outweighed by the benefit of allowing dapps to update before the switch to PoS so as soon as PoS lands they can adopt safer practices.

One option would be to allow individual Ethereum clients to configure what `finalized` and `safe` mean prior to The Merge, and then post merge these config options would be ignored.  This would allow an exchange, for example, to set `finalized` to be `latest - 6` (if that is what they want).  We can then have "reasonable defaults" for people who don't manually configure, which err on the side of safety (like `finalized = latest -256`).

mkalinin replied:
> My problem with waiting until post merge to add these is that it leaves us in a situation where dapps cannot deploy merge-ready prior to sometime *after* the merge.

Dapps may fallback to k-confirmations rule if `finalized` has been responded with error -- this is one of the ways to get prepared pre-Merge. This may not work well for the transition period, when no execution block is yet finalized and network has left PoW. But I would expect dapps consider blockchain reads slightly prior and post Merge as potentially insecure and stop processing deposits and withdrawals for this period of time.

We may have `safe == justified_block_hash`. At least safe head algorithm starts with justified, and justified is usually an epoch closer to the head than `finalized`. Once `get_safe_head` is fully implemented, UX of the `safe` will be improved as this block will become even closer to the head.

mkalinin replied:
Also, I am not sure about "unsafe". If we take justified as a safe block path then "unsafe" becomes an alias to "latest" - not sure if "usnafe" will ever be used unless we deprecate "latest". It should be said that using "latest" is unsafe/unreliable for most of the cases and a lot of care should be taken for one that wants to use it but not sure if "unsafe" should ever replace "latest" because the latter has different semantics meaning (the tip of the canonical chain at the moment) which is getting lost if replaced with "unsafe".

MicahZoltu replied:
I do think `latest` should be deprecated and eventually (an undefined point in the distant future) removed.

Why do you think the semantic meaning is lost with `unsafe`?

mkalinin commented:
"safe" and "unsafe", and the description are added. This PR is ready for review now. I will create a separate one to add "justified".

MicahZoltu commented on src/schemas/block.json:
```diff
@@ -125,7 +125,8 @@
 		"type": "string",
-		"enum": ["earliest", "latest", "pending"]
+		"enum": ["earliest", "latest", "pending", "finalized", "safe", "unsafe"],
+		"description": "Tags \"finalized\" and \"safe\" refer to the corresponding blocks defined by the most recent fork choice state. Tag \"unsafe\" is an alias to the \"latest\"."
```
Is there a compelling reason to use double quotes here instead of single quotes?  The raw JSON is easier to read with single quotes (no need to escape) and should work just as well.
```suggestion
		"description": "Tags 'finalized' and 'safe' refer to the corresponding blocks defined by the most recent fork choice state. Tag 'unsafe' is an alias to the 'latest'."
```

MicahZoltu commented on src/schemas/block.json:
```diff
@@ -125,7 +125,8 @@
 		"title": "Block tag",
 		"type": "string",
-		"enum": ["earliest", "latest", "pending"]
+		"enum": ["earliest", "latest", "pending", "finalized", "safe", "unsafe"],
+		"description": "Tags \"finalized\" and \"safe\" refer to the corresponding blocks defined by the most recent fork choice state. Tag \"unsafe\" is an alias to the \"latest\"."
```
Are newlines supported in the description string?  If not we could replace the `\n` below with `; ` or something.  I think there is value in defining each fairly clearly.
```suggestion
		"enum": ["earliest", "finalized", "safe", "unsafe", "pending"],
		"description": "`earliest`: The lowest numbered block the client has available.\n`finalized`: The most recent block that cannot be reorged outside of a hard fork.\n`safe`: The most recent block that is will not reorg unless there is an active ongoing attack.\n`unsafe`: The most recent block the client has seen that it believes will be part of the canonical chain (high probability of being reorged out).\n`pending`: A sample next block built by the client containing what it believes will be the set of transactions likely to show up in that block.\n`latest`: **DEPRECATED** Will be removed at some point in the future, currently an alias for `unsafe` or `safe`."
```
I also removed `latest` from the enum and only referenced it in the description to help make it clear that it is deprecated (but if people go looking for it they can find it).

mkalinin replied:
If we remove `latest` it would read as it's already unaccessible. I can foresee users not finishing reading the description (thus, not seeing that it's been DEPRECATED but still available for a short period of time) and starting to ask why this tag that was in place for almost 7 years has gone.

For the same reason I think if we leave it there nobody will read that it's deprecated, thus, is about to be removed soon and shouldn't be used. 🤔

mkalinin replied:
Newlines aren't supported and escaped with a space, "; " should work.

MicahZoltu replied:
> If we remove `latest` it would read as it's already unaccessible. I can foresee users not finishing reading the description (thus, not seeing that it's been DEPRECATED but still available for a short period of time) and starting to ask why this tag that was in place for almost 7 years has gone.
> 
> For the same reason I think if we leave it there nobody will read that it's deprecated, thus, is about to be removed soon and shouldn't be used. 🤔

I think I would rather have people asking around when they failed to read rather than having people continuing to use the deprecated thing when they failed to read.

mkalinin replied:
I hope developers will not remove it faster than we expect this to happen. My concern is that there is no other place where the set of values for this enum is specified. If this document were for users only, and the spec would still contain `latest` (marked as deprecated) then I would be less concerned.

MicahZoltu reviewed (approved):
I still have comments, but none that should block this from being merged.

MicahZoltu commented on src/schemas/block.json:
```diff
@@ -125,7 +125,8 @@
 		"type": "string",
-		"enum": ["earliest", "latest", "pending"]
+		"enum": ["earliest", "finalized", "safe", "unsafe", "pending"],
+		"description": "`earliest`: The lowest numbered block the client has available.\n`finalized`: The most recent block that cannot be reorged outside of a hard fork.\n`safe`: The most recent block that is will not reorg unless there is an active ongoing attack.\n`unsafe`: The most recent block the client has seen that it believes will be part of the canonical chain (high probability of being reorg ...
```
Removes newlines and uses `;` instead (newlines apparently aren't supported in OpenRPC).

```suggestion
		"description": "`earliest`: The lowest numbered block the client has available; `finalized`: The most recent block that cannot be reorged outside of a hard fork; `safe`: The most recent block that is will not reorg unless there is an active ongoing attack; `unsafe`: The most recent block the client has seen that it believes will be part of the canonical chain (high probability of being reorged out); `pending`: A sample next block built by the client containing what it believes will be the set of transactions likely to show up in that block; `latest`: **DEPRECATED** Will be removed at some point in the future, currently an alias for `unsafe` or `safe`."
```

mkalinin commented:
I have taken descriptions suggested by @MicahZoltu and modified them a bit in this document https://hackmd.io/T8LBBP6HSLqGFWM8ikQakA. Let's agree on the texts in hackmd and then just move it into the schema. I feel it should be more convenient to do

lightclient commented:
Theoretically, the `description` field [should support](https://spec.open-rpc.org/#rich-text-formatting) Github markdown.

MicahZoltu commented:
> Theoretically, the `description` field [should support](https://spec.open-rpc.org/#rich-text-formatting) Github markdown.

I believe the problem is that JSON doesn't allow newlines, and Markdown doesn't have a mechanism for inserting newlines other than with the newline character (just as it doesn't have a mechanism for expressing an `a` other than by using the `a` character).  This means that there is functionally no way to get a newline inserted, even though it would be supported if you could get it in there.

MicahZoltu commented:
> We should probably make a decision on this before merging?

It will alias to `unsafe` for now, but we may move it to `safe` in the future.

djrtwo reviewed (commented):
I'm weakly against removing `latest` from the enum at this point given we don't have a good `safe` algorithm that follows near the head. thus steering users to use `safe` for most purposes doesn't yet capture the intent -- it trails by 6 to 10 minutes in current "justified" algorithm.

My preference would be to deprecate `latest` once we have a good (near head in normal case) `safe`  algorithm. Until then, I feel that `latest` has high semantic value.

> high probability of being reorged out

also, I think this is a bit extreme of language when we have the huge spread between safe and unsafe today. a block a few blocks deep in most cases does not have a high probability of being re-org'd out.

If we move forward as is, I'll be strongly against *fully* removing `latest` until the safe/unsafe alg disparity is fixed

djrtwo commented:
> Is an attack really only the possibility of reorging a safe block?

@lightclient in the current algorithm which only uses "justified", then yes -- an attack or a bug.

The assumptions we'll likely use in a future "safe" algorithm will have a couple more assumptions -- 50%+ honest (non-attacking) and some synchrony bound. So it's not just an attack, a high asynchrony (depending on the assumption in the algorithm) could also re-org blocks

mkalinin commented:
I have updated the description according to [the doc](https://hackmd.io/T8LBBP6HSLqGFWM8ikQakA).

> My preference would be to deprecate `latest` once we have a good (near head in normal case) safe algorithm. Until then, I feel that `latest` has high semantic value.

I concur. `latest` is currently marked as deprecated and it is said that it will be removed in the future -- i think this is pretty visible announcement taking in account that users will read the description because of new tags.

djrtwo commented:
I've been thinking about this, and I actually don't think `unsafe` is correct/valuable word here even in the long run (or especially in the long run).

In the event that we define a safe head algorithm that more often than not *is* latest (is the head), then in such a case, "safe" and "unsafe" would often return the same thing. This is semantically unsound and confusion -- something that is safe cannot simultaneously be unsafe.

Thus I think it is actually best to keep "latest" -- a word that has strong semantic meaning and (I argue) will always be of value to some set of users. And add "safe" (an algorithm we hope to improve) and 'finalized"

That is -- do not deprecate "latest" and do not add "unsafe"

dankrad commented:
I don't think I agree with this interpretation of the semantics. "safe" and "unsafe" are not referring to a block, but instead to the way that block was derived. Both methods can indeed come to the same block, but this is not a contradiction.

mkalinin commented:
> I don't think I agree with this interpretation of the semantics. "safe" and "unsafe" are not referring to a block, but instead to the way that block was derived. Both methods can indeed come to the same block, but this is not a contradiction.

I don't think that users associate these tags with the way of obtaining the blocks, it could be too involving for an average user. These tags will highly likely be associated with the blocks itself, thus, `safe` and `unsafe` referring to the same block is gonna be a source of confusion.

djrtwo commented:
Agreed that you can interpret these as "the algorithm used to derive" but as Mikhail mentioned, I think it is incredibly unlikely that normal users (and even sophisticated ones at first glance!) would assume and properly interpret this.

MicahZoltu commented:
I think one could make the same argument that users won't think through or even notice in most cases that "safe and unsafe" are the same.  They'll likely try to understand at a very high level what each means, pick one, and query it (using whatever result it returns).

mkalinin commented:
I've added "Before the merge transition is finalized any call querying over `finalized` or `safe` block MUST be responded with error" to the end of description -- looks ugly but this seems like the best we can do. I would rather specify an exact error code and message but apparently there is no unified list of errors for JSON-RPC and it may vary depending on client which means that whatever code we pick up we're risking to get into a conflict with already existing one or the one that will be introduced later on. If anyone have any ideas on figuring out the code we may use for this purpose I am keen to discuss

MicahZoltu commented:
I recommend just picking an error code and message that seems reasonable and proposing it.  Tell the execution clients, and if they find it overlaps with something they have we can change it.  I think having no standard error code/message is far worse than having a conflicting error code/message.

mkalinin commented:
> I recommend just picking an error code and message that seems reasonable and proposing it. Tell the execution clients, and if they find it overlaps with something they have we can change it. I think having no standard error code/message is far worse than having a conflicting error code/message.

I picked up arbitrarily large custom error code. It should not conflict with already existing ones, and if we hit this code at some point in time we should be pretty far from the Merge which should preclude potential error code conflicts.

MicahZoltu commented:
As long as it follows the constraints here, I don't care what you pick.  😁 
https://www.jsonrpc.org/specification#error_object

mkalinin commented:
> As long as it follows the constraints here, I don't care what you pick. 😁
> https://www.jsonrpc.org/specification#error_object

It's set to `-32100` which is within the range of application defined errors

MicahZoltu commented:
I believe application defined errors are "everything outside of `[-32000,-32768]`".  `-32100` would be for "server defined errors" (I don't know what that means).
> -32000 to -32099 | Server error | Reserved for implementation-defined server-errors.
> The remainder of the space is available for application defined errors.

mkalinin commented:
> I believe application defined errors are "everything outside of `[-32000,-32768]`". `-32100` would be for "server defined errors" (I don't know what that means).
> 
> > -32000 to -32099 | Server error | Reserved for implementation-defined server-errors.
> > The remainder of the space is available for application defined errors.

Yes, you're right. And it looks like we have messed up with this range in Engine API as well

mkalinin commented:
I've replaced `unsafe` with `latest` and removed DEPRECATED note from the latter. Please, take a look once again and if everything is OK let's get it merged.

MicahZoltu commented on src/schemas/block.json:
```diff
@@ -125,7 +125,8 @@
 		"type": "string",
-		"enum": ["earliest", "latest", "pending"]
+		"enum": ["earliest", "finalized", "safe", "latest", "pending"],
+		"description": "`earliest`: The lowest numbered block the client has available; `finalized`: The most recent crypto-economically secure block, cannot be re-orged outside of manual intervention driven by community coordination; `safe`: The most recent block that is safe from re-orgs under honest majority and certain synchronicity assumptions; `latest`: The most recent block in the canonical chai ...
```
```suggestion
		"description": "`earliest`: The lowest numbered block the client has available; `finalized`: The most recent crypto-economically secure block, cannot be re-orged outside of manual intervention driven by community coordination; `safe`: The most recent block that is safe from re-orgs under honest majority and certain synchronicity assumptions; `latest`: The most recent block in the canonical chain observed by the client, this block may be re-orged out of the canonical chain even under healthy/normal conditions; `pending`: A sample next block built by the client on top of `latest` and containing the set of transactions usually taken from local mempool. Before the merge transition is finalized, any call querying for `finalized` or `safe` block MUST be responded to with `-39001: Unknown block` error"
```

MicahZoltu reviewed (approved):
Just some grammar stuff and a minor addition to the definition of `latest`.

mkalinin commented:
> Just some grammar stuff and a minor addition to the definition of `latest`.

Applied
